### PR TITLE
Initial design for IngressRoute/v1 routing

### LINKED ISSUE #947: Header based routing
Expose header based routing https://www.envoyproxy.io/docs/envoy/v1.5.0/api-v1/route_config/route#headers

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #947, updates #1075 by providing an initial design for routing in IngressRoute/v1. The goal of this design is to determine how we can further enable routing in Contour but still maintain proper delegation & multi-team capabilities. 

Signed-off-by: Steve Sloka slokas@vmware.com

### CODE REVIEW

File: design/routing-design.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    services:
+    - name: glob-service
+      port: 80
+```
+
+### Virtual Host:
+
+Contour today only supports exact domain name matching when defining the `fqdn` in an IngressRoute. However, supporting additional types is possible since the virtual host level doesn't require the same level of detail in regards to delegation. 
+
+Contour wil support the same [domain search order](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto#route-virtualhost) as Envoy. Given this order, IngressRoutes that are defined will follow this logic. Contour should also review this and mark the status of IngressRoutes accordingly:
```
[the remark this answers is not in the document]
Comment (timh) [reply design/routing-design.md]: Is it limiting to tie our domain search ordering to Envoy's standard? Is the standard good? Does this tie us to Envoy at the design level in a way that's acceptable?

File: design/routing-design.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+  routes:
+  - match:
+    path: /foo
+    - header: x-header
+      value: a
+    delegate:
+      name: headera-ir
+      namespace: team-a
+  - match: 
+      path: /foo
```
Comment (davecheney) [inline design/routing-design.md]: The indentation here looks off, is this intentional, or did github eat it?

File: design/routing-design.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevesloka) [reply design/routing-design.md]: The ordering is more of an implementation detail, I don't think it affects the api design.

**Comment by stevesloka** [comment]
@dfc here are some replies to your comments, the others I'm going to address directly in the document. I'll address these comments as well, but some are just discussion: 

> 2. Types of routing. This document identifies header-based routing, exact path, and regex as new additions. The document outlines the use cases for these types but does not explain why they were chosen. We need to have an answer when people come asking for "can I route by source IP?", for example. Maybe the answer is "we can't do it", or maybe just "we haven't done it yet", but we need to have a justification.

I thought the use-cases were the reasons we chose those types of routing. So given users would want to solve these types of problems (i.e. the use-cases), we chose those types. I don’t have a good use-case for routing via source IP but do see for rate-limiting or allow/disallow lists by source IP. Do you want a section to call out types that we won't’ support? For example, we could add regex to this list or JWT tokens? 

> 4. Exact path routing. How does this interact with delegation? Can I delegate exact paths or only prefixes?


You can still delegate exact paths, but they won’t prefix match. For example, if you had `/foo/` then you could delegate `/foo/bar/` such that only `/foo/` would match, `/foo/bar/` would match, but `/foo/barbaz` would not. 

> 6. In the high-level design, it states that regex is not supported except in delegation. What is the reason for this restriction (other than my concern that it's not possible)? What does that mean when one ingressroute delegates to another? Is regex only supported in leaves, not intermediaries? How can we establish those business rules? How can we communicate them to users when they are confused?

My thought was to allow for Regex but put restrictions on how you can use them. If you want to use, then they must be at the end of a delegation chain. So in a sense, we don’t support delegation with Regex, it’s just the final path that you can create.

> 7. From the design document, it seems that delegation forces us into a model of prefix plus another routing type; eg. prefix plus a header.

Yes, there isn’t a way to route with Envoy without a path. At one point I did have some thoughts around allowing users of Contour specify ONLY a header, Contour would then write in a regex path to cover the missing path that the user did not supply. For example, if you specify a host header to route to a specific backend, say iOS vs Android users. By specifying only the header, Contour would write `/*` into the path and match anything. Then you could specify a path `/foo` that should then I guess have the header specifics applied to all child delegations, otherwise, you would need to specify the header on each route down the chain. 

> 8. Header routing. If the routing model is prefix plus something, is it prefix plus a header? What is the justification for making header matching a substring? What about if people want to route on the presence of the header, or the absence of it, or if (shudder) they want to use regex to match a header value? I don't think what's written is incorrect, but it needs more justification.

Header routing must be regex since a common header to match in is a user-agent string. We could match on the presence of a header by not specifying a `value`. I’m not sure about how to specify routing on the absence of a header. Do you have any thoughts on how this could look?

**Comment by stevesloka** [comment]
I just pushed another big rev of the design. One thing I'm bouncing in my head is the term "glob". I switched to the term `wildcard` but that has other meanings, so maybe should switch back to glob. My initial thought was that `glob` may not be clear what it's doing.

**Comment by davecheney** [comment]
I'm going to commit this as a draft as is so we capture the discussion.

**Comment by glerchundi** [comment]
@stevesloka as far as i can see the current design doesn't capture the possibility to route on the mere fact of the existence of a specific header regardless of the value of that one. Is this on purpose?

Something like this:

```
apiVersion: projectcontour.io/v1
kind: IngressRoute
metadata: 
  name: example
spec:
  routes:
  - match:
      path: /foo
      - header: x-header
    services:
    - name: backend-a
      port: 9999
```

Non-existing value key means it ignores the value and checks only if the header is present.

**Comment by stevesloka** [comment]
@glerchundi yup we're planning on having that feature. I have some additional examples that will help which I need to find the time and update this document further with those examples.

**Comment by glerchundi** [comment]
Thanks Steve, that’s great. Closely following the hard work being done here
👏🏻

**Pull request merged**